readme.md
GetMyComponent Extra for MODx Revolution
=======================================


**Author:** Bob Ray <http://bobsguides.com> [Bob's Guides](http://bobsguides.com)

Documentation is available at [Bob's Guides](http://bobsguides.com/mycomponent-tutorial.html)

GetMyComponent creates the files for the MyComponent package, a generic build script for MODx Revolution components.
Since MyComponent is itself a build script, it can't be installed through Package Manager. Once you have
installed GetMyComponent, the MyComponent files will be in place at assets/mycomponents/mycomponent/.

From that point on, the GetMyComponent package should be ignored. Note that if you uninstall
or remove it, all the files of MyComponent will be removed.


docs/changelog.txt
Changelog for GetMyComponent

GetMyComponent 1.0.0
-------------------
Initial Version

_build/preinstall.script.php
<?php
/**
 * GetMyComponent pre-install script
 *
 * Copyright 2011 Your Name <you@yourdomain.com>
 * @author Your Name <you@yourdomain.com>
 * 1/1/11
 *
 * GetMyComponent is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option) any
 * later version.
 *
 * GetMyComponent is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * GetMyComponent; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * @package GetMyComponent
 */
/**
 * Description: Attempts to create assets/components/GetMyComponents/GetMyComponent
 * directory.
 *
 * @package GetMyComponent
 * @subpackage build
 */
/**
 * @package GetMyComponent
 * Validators execute before the package is installed. If they return
 * false, the package install is aborted. This example attempts to
 * create the assets/components/mycomponents/mycomponent directory
  * and aborts the install if it is not found.
 */

/* The $modx object is not available here. In its place we
 * use $object->xpdo
 */
$modx =& $object->xpdo;


$modx->log(xPDO::LOG_LEVEL_INFO,'Running PHP Validator.');
switch($options[xPDOTransport::PACKAGE_ACTION]) {
    case xPDOTransport::ACTION_INSTALL:

        $modx->log(xPDO::LOG_LEVEL_INFO,'Attempting to create assets/components/mycomponents directory.'); flush();
        $success = true;
        $outfile = MODX_ASSETS_PATH . 'components/mycomponents/';
        if (file_exists($outfile) == false) {
            $makeDir = mkdir($outfile, 0700, true);
            if (! $makeDir) {
                $modx->log(xPDO::LOG_LEVEL_INFO,'Failed to create directory ' . $outfile . '. . . Create it yourself and reinstall.<br />');flush();
                $success = false;
            }
        }
        break;
   /* These cases must return true or the upgrade/uninstall will be cancelled */
   case xPDOTransport::ACTION_UPGRADE:
        $success = true;
        break;

    case xPDOTransport::ACTION_UNINSTALL:
        $success = true;
        break;
}

return $success;

_build/build.transport.php
<?php
/**
 * GetMyComponent Build Script
 *
 * Copyright 2011 Bob Ray <http://bobsguides.com>
 *
 * GetMyComponent is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option) any
 * later version.
 *
 * GetMyComponent is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * GetMyComponent; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * @package GetMyComponent
 * @subpackage build
 */
/**
 * GetMyComponent Package
 *
 * Description: Installs the files of the GetMyComponent build
 * script template.
 *
 * @package GetMyComponent
 * @subpackage build
 */


/* See the GetMyComponent/core/docs/notes.txt file for
 * more detailed information about using the package
 *
 * Search and replace tasks:
 * (edit the resource and element names first if they have
 * different names than your package.)
 *
 * GetMyComponent -> Name of your package
 * GetMyComponent -> lowercase name of your package
 * Your Name -> Your Actual Name
 * Your Site -> Name of your site
 * yoursite -> domain name of your site
 * you@yourdomain.com -> your email address
 * Description -> Description of file or component
 *
 * 1/1/11 -> Current date
 * 2011 -> Current Year
 */

/* Set package info be sure to set all of these */
define('PKG_NAME','GetMyComponent');
define('PKG_NAME_LOWER','getmycomponent');
define('PKG_VERSION','1.0.0');
define('PKG_RELEASE','beta1');



/******************************************
 * Work begins here
 * ****************************************/

/* set start time */
$mtime = microtime();
$mtime = explode(" ", $mtime);
$mtime = $mtime[1] + $mtime[0];
$tstart = $mtime;
set_time_limit(0);

/* define sources */
$root = dirname(dirname(__FILE__)) . '/';

$sources= array (
    'root' => $root,
    'build' => $root . '_build/',
    'docs' => $root . 'docs/',
);
unset($root);
$sourcePath = dirname(dirname(dirname(__FILE__))) . '/mycomponent';
$targetPath = "return MODX_ASSETS_PATH . 'mycomponents/';";

/* Instantiate MODx -- if this require fails, check your
 * _build/build.config.php file
 */
require_once $sources['build'].'build.config.php';
require_once MODX_CORE_PATH . 'model/modx/modx.class.php';
$modx= new modX();
$modx->initialize('mgr');

$modx->setLogLevel(xPDO::LOG_LEVEL_INFO);
$modx->setLogTarget(XPDO_CLI_MODE ? 'ECHO' : 'HTML');
$modx->log(xPDO::LOG_LEVEL_INFO, "Initializing MODx."); flush();
$modx->log(xPDO::LOG_LEVEL_INFO, "Loading Package Builder."); flush();
/* load builder */
$modx->loadClass('transport.modPackageBuilder','',false, true);
$builder = new modPackageBuilder($modx);
$package = $builder->createPackage(PKG_NAME_LOWER, PKG_VERSION, PKG_RELEASE);
$builder->registerNamespace(PKG_NAME_LOWER,false,true,'{core_path}components/'.PKG_NAME_LOWER.'/');
flush();

$modx->log(xPDO::LOG_LEVEL_INFO, "Packaging Files (this may take a while) . . ."); flush();

$obj = array ('source' => $sourcePath,'target' =>$targetPath);
$attributes = array('vehicle_class' => 'xPDOFileVehicle');
$package->put($obj,$attributes);

$builder->setPackageAttributes(array(
    'license' => file_get_contents($sources['docs'] . 'license.txt'),
    'readme' => file_get_contents($sources['docs'] . 'readme.txt'),
    'changelog' => file_get_contents($sources['docs'] . 'changelog.txt'),

));
$modx->log(xPDO::LOG_LEVEL_INFO, "Zipping up Package. Please wait . . ."); flush();

/* Last step - zip up the package */
$builder->pack();

/* report how long it took */
$mtime= microtime();
$mtime= explode(" ", $mtime);
$mtime= $mtime[1] + $mtime[0];
$tend= $mtime;
$totalTime= ($tend - $tstart);
$totalTime= sprintf("%2.4f s", $totalTime);

$modx->log(xPDO::LOG_LEVEL_INFO, "Package Built.");
$modx->log(xPDO::LOG_LEVEL_INFO, "Execution time: {$totalTime}");
exit();
